lib/services/product-manager.js

```diff
@@ -415,4 +416,88 @@ export default class ProductManager {
   async _unpublishProduct (product) {
     return this.updateProduct(product, [this._getUnpublishAction()])
   }
+
+  /**
+   * Ensure that all attributes that have sameForAll constraint will have same value after
+   * the reassignment process. The same value is taken either from product draft OR set to null
+   * (attr is removed) if attr is not available in product draft.
+   * It's necessary to have same attribute values for:
+   *
+   * 1) variants that will be added to ctpProductToUpdate - done by removing the attribute
+   * from the new variants
+   *
+   * 2) existing variants from ctpProductToUpdate - done by `setAttributeInAllVariants` action
+   *
+   * @param variants - checks if the values for sameForAll attributes are correct in the variants
+   * @param productTypeId - fetch the product type by ID and get all sameForAll attrs for checking
+   * @param productDraft - get sameForAll value from the productDraft
+   * in case of sameForAll constraint violation
+   * @returns {Promise.<Array>} setAttributeInAllVariants actions
+   */
+  async ensureSameForAllAttributes (variants, productTypeId, productDraft) {
+    const actions = []
+    const productType = await this._getProductTypeById(productTypeId)
+    const sameForAllAttrs = this._selectSameForAllAttrs(productType)
+    const violatedAttrs = this._selectViolatedAttrs(variants, sameForAllAttrs)
+    violatedAttrs.forEach((attribute) => {
+      const value = this._getAttributeValue(attribute.name, productDraft)
+      variants.forEach((variant) => {
+        const attrFromVariant = variant.attributes.find(a => a.name === attribute.name)
+        if (value)
```

Comment:
is it necessary to update local version of product variant attributes?

Reply:
The idea is this method can accept:
1) variants from new product draft. Therefore I update the value from the attribute, because the product draft will then be created.
2) variants from existing product. In this case, it's necessary to create action `setAttributeInAllVariants`.

That's why I update the variant and also at the same time return the action. And the caller decides, what to do next.